aws-cpp-sdk-core-tests/utils/threading/ExecutorTest.cpp

```diff
@@ -0,0 +1,373 @@
+            if (lastTaskWaiting)
+            {
+                break;
+            }
+        }
+    }
+    taskSignal.notify_all();
+            
+    locker.lock();
+    
+    if (taskOrder.size() != FINAL_SIZE)
+    {
+        doneSignal.wait(locker, [&] {
+            return taskOrder.size() == FINAL_SIZE;
+        });
+    }
+    
+    // Check completion and ordering of tasks.
+    ASSERT_EQ(taskOrder.size(), FINAL_SIZE);
+    ASSERT_EQ(taskOrder.at(3), 3ul);
+    locker.unlock();
+    
+    AWS_END_MEMORY_TEST
+}
+
+/** Tests whether the blocking executor properly delegates tasks concurrently */
+TEST(BlockingExecutorTest, ConcurrencyTimingTest)
+{
+    AWS_BEGIN_MEMORY_TEST(16, 10)
+    
+    const size_t NUM_TASKS = 15;
+    const size_t POOL_SIZE = 5;
+    
+    unsigned int counter = 0;
+    std::mutex counterLock;
+    std::condition_variable doneSignal;
+    
+    auto executeTask = [&] () {
+        std::this_thread::sleep_for(milliseconds(10));
+        std::unique_lock<std::mutex> locker(counterLock);
+        counter++;
+        if (counter == NUM_TASKS)
+        {
+            locker.unlock();
+            doneSignal.notify_all();
+        }
+    };
+
+    // Time how long it takes to run the tasks with an underlying executor that's concurrent
+    std::shared_ptr<DefaultExecutor> mockExecutor =
+            Aws::MakeShared<DefaultExecutor>(EXECUTOR_TEST_ALLOCATION_TAG);
+    BlockingExecutor blockingExecutor(mockExecutor, POOL_SIZE);
+    
+    high_resolution_clock::time_point startTime = high_resolution_clock::now();
+    for (u_long i = 0; i < NUM_TASKS; i++)
+    {
+        blockingExecutor.Submit(executeTask);
+    }
+    std::unique_lock<std::mutex> locker(counterLock);
+    while (counter != NUM_TASKS)
+    {
+        doneSignal.wait(locker);
+    }
+    counter = 0;
+    locker.unlock();
+    high_resolution_clock::time_point endTime = high_resolution_clock::now();
+    auto asyncDuration = duration_cast<milliseconds>(endTime - startTime).count();
+    
+    // Time how long it takes to run the tasks with an underlying executor that doesn't use threads
+    std::shared_ptr<ThreadlessMockExecutor> threadlessMockExecutor =
+            Aws::MakeShared<ThreadlessMockExecutor>(EXECUTOR_TEST_ALLOCATION_TAG);
+    BlockingExecutor threadlessBlockingExecutor(threadlessMockExecutor, POOL_SIZE);
+
+    startTime = high_resolution_clock::now();
+    for (u_long i = 0; i < NUM_TASKS; i++)
+    {
+        threadlessBlockingExecutor.Submit(executeTask);
+    }
+    locker.lock();
+    while (counter != NUM_TASKS)
+    {
+        doneSignal.wait(locker);
+    }
+    endTime = high_resolution_clock::now();
+    auto sequentialDuration = duration_cast<milliseconds>(endTime - startTime).count();
+    
+    /* Check if the concurrent time was faster, with an adjustable tolerance for the strictness of
+     * the test */
+    const double TOLERANCE = 1.0;
+    ASSERT_LE(asyncDuration, sequentialDuration * TOLERANCE);
```

Comment:
I was also concerned about this but didn't know how else to code this.
The test is supposed to assert that when you allow a lot of parallel threads, things actually run in parallel. The time for each task (10 ms -- we could make it 100) is a lot longer than anything else that runs on the CPU on any system I know. It's O(I/O time). So this should be perfectly safe. The only scary thing, I think, is the inaccuracy of the time measurement but again, it shouldn't be anything comparable to ms.
We run 15 tasks and I did assume we will have at least 2 cores... otherwise well obviously you're running serially. Maybe we can run this test only if we have >= 2 cores ?

I am also open to any other suggestions or just deleting this test. It did its job as a unit test.